### DISCUSSION TITLE
Add option to monitor keystrokes and suggest expansions à la TextExpander

### DISCUSSION THREAD

**Discussion by seanlaidlaw** [espanso/espanso#1210] [Feature requests and Ideas]
Are there plans to introduce a feature like in TextExpander where monitoring of non-secure keystrokes can be enabled, to obtain suggestions of useful expansions ? This is the main thing I miss from being a  TextExpander user

**Comment by federico-terzi** [comment]
Hey @seanlaidlaw,

Thank you for the feedback! I've been thinking about this feature for a long while, as I would find it very useful as well. The biggest problem comes from the "*where monitoring of non-secure keystrokes can be enabled*" part, as that's not trivial to implement.

While on macOS the OS enforces the `SecureInput` policy that automatically prevents espanso from listening from secure keystrokes, the same does not happen on Windows or Linux. I've made some tests, and I haven't found a way to reliably detect whether the user is currently typing a password or not on such systems.

This is not a problem with the current version of espanso, as it doesn't save the typed characters anywhere and only keeps them in memory for a few seconds, but if we enable this kind of feature, espanso would somehow need to "save" those sentences to later suggest them.

As you can imagine, from a security standpoint this would be a very dangerous situation, and therefore I decided to avoid it altogether. I'm not excluding to implement it if a secure implementation is suggested, but I still haven't found it :)

Cheers :)

**Comment by ghost** [comment]
@federico-terzi how about we have a pattern to non-security sensitive matches to do suggestions. I think suggestions is a very cool feature.

**Comment by federico-terzi** [comment]
@rhacker Thank you for the feedback! What kind of patterns do you have in mind?
